src/new_client/connection.rs

```diff
@@ -0,0 +1,218 @@
+        let (sub_sender, sub_receiver) = mpsc::unbounded();
+
+        Ok(Connection {
+            thread: thread::spawn(move || {
+                // TODO(stjepang): Report errors from `run()` in a better place.
+                dbg!(smol::run(client(&url, pub_receiver, sub_receiver)))
+            }),
+            sub_ops: sub_sender,
+            pub_ops: pub_sender,
+            sid_gen: AtomicUsize::new(1),
+        })
+    }
+
+    /// Publishes a message.
+    pub fn publish(&mut self, subject: &str, msg: impl AsRef<[u8]>) -> io::Result<()> {
+        let subject = subject.to_string();
+        let msg = msg.as_ref().to_vec();
+
+        // Enqueue a PUB operation.
+        self.pub_ops
+            .send((subject, msg))
+            .now_or_never()
+            .expect("future can't be pending because the publish channel is unbounded")
+            .expect("the publish channel shouldn't be disconnected");
+
+        Ok(())
+    }
+
+    /// Creates a new subscriber.
+    pub fn subscribe(&mut self, subject: &str) -> Subscription {
+        let sid = self.sid_gen.fetch_add(1, Ordering::SeqCst);
+        Subscription::new(subject, sid, self.sub_ops.clone())
+    }
+}
+
+impl Drop for Connection {
+    fn drop(&mut self) {
+        // TODO(stjepang): Wait for `self.thread` to stop.
+    }
+}
+
+/// The main loop for a NATS client.
+async fn client(
+    url: &str,
+    mut pub_ops: mpsc::UnboundedReceiver<(String, Vec<u8>)>,
+    mut sub_ops: mpsc::UnboundedReceiver<SubscriptionOp>,
+) -> io::Result<()> {
+    let stream = Arc::new(Async::<TcpStream>::connect(url).await?);
+    let mut writer = stream.clone();
+
+    // Create an endless stream parsing operations from the server.
+    let mut reader = stream::try_unfold(BufReader::new(stream), |mut stream| async {
+        // Decode a single operation.
+        let op = decode(&mut stream).await?;
+        io::Result::Ok(Some((op, stream)))
+    })
+    .boxed();
+
+    // Expect an INFO message.
+    let server_info = match reader
+        .try_next()
+        .await?
+        .expect("end of what should be an infinite stream")
+    {
+        ServerOp::Info(server_info) => server_info,
+        _ => return Err(Error::new(ErrorKind::Other, "expected an INFO message")),
+    };
+
+    // Send a CONNECT operation to the server.
+    encode(
+        &mut writer,
+        ClientOp::Connect(ConnectInfo {
+            tls_required: false,
+            name: None,
+            pedantic: false,
+            verbose: false,
+            lang: crate::LANG.to_string(),
+            version: crate::VERSION.to_string(),
+            user: None,
+            pass: None,
+            auth_token: None,
+            user_jwt: None,
+            signature: None,
+            echo: true,
+        }),
+    )
+    .await?;
+
+    // Current subscriptions in the form `(subject, sid, messages)`.
+    let mut subscriptions: Vec<(String, usize, mpsc::UnboundedSender<Message>)> = Vec::new();
```

Comment:
No locks on subscriptions, the whole client state is contained inside this `client()` function :)